// github-pages/build/lib/parse-config.js
'use strict';

var pkgConf = require('pkg-conf');
var objectMerge = require('./object-merge');

var defaultCfg = {
  api: {
    version: '3.0.0',
    protocol: 'https',
    host: 'api.github.com',
    pathPrefix: '',
    timeout: 5000
  },
  auth: {
    type: 'token',
    token: process.env.GH_TOKEN
  },
  remote: {
    ref: 'heads/gh-pages'
  },
  commit: {
    message: 'github-pages publish.'
  }
};

function parseRepo(repoSlug) {
  var slugExp = /^([^\/]+)\/([^\/]+)$/;
  var slug = slugExp.exec(repoSlug);
  if (!slug) {
    return {};
  }

  return {
    user: slug[1],
    repo: slug[2]
  };
}

function parseAuthor(author) {
  var emailExp = /^.*<(.+)>.*$/;
  var nameExp = /^(.*\S).*<.*$/;

  try {
    return {
      name: nameExp.exec(author)[1],
      email: emailExp.exec(author)[1]
    };
  } catch (e) {
    return undefined;
  }
}

function flagsToCfg(flags, src) {
  var _parseRepo = parseRepo(flags.repo);

  var user = _parseRepo.user;
  var repo = _parseRepo.repo;

  var author = parseAuthor(flags.commitAuthor);

  return {
    api: {
      version: flags.apiVersion,
      protocol: flags.apiProtocol,
      host: flags.apiHost,
      pathPrefix: flags.apiPath,
      timeout: flags.apiTimeout
    },
    auth: {
      type: flags.token ? 'token' : undefined,
      token: flags.token
    },
    commit: {
      message: flags.commitMessage,
      author: author
    },
    remote: {
      user: user,
      repo: repo,
      ref: flags.remoteRef
    },
    src: src && src.length > 0 ? src : undefined
  };
}

function isValid(config) {
  var _ref = config || {};

  var remote = _ref.remote;
  var auth = _ref.auth;
  var src = _ref.src;

  return remote && remote.repo && remote.user && auth && auth.type && auth.token && src && src.length > 0;
}

module.exports = function (flags, src) {
  var fileCfg = pkgConf.sync('github-pages');
  var cliCfg = flagsToCfg(flags, src);
  var config = objectMerge(defaultCfg, fileCfg, cliCfg);

  if (!isValid(config)) {
    return null;
  }

  return config;
};

module.exports.default = defaultCfg;
module.exports.isValid = isValid;
